pkg/agent/discoverycache.go

```diff
@@ -24,264 +25,428 @@ const (
+		discoveryFuncs = append(discoveryFuncs, dc.additionalDiscoveryFuncs...)
 	}
 
-	agent.cacheManager.SaveCache()
-	j.logger.Trace("cache has been updated and saved")
+	errCh := make(chan error, len(discoveryFuncs))
+	wg := &sync.WaitGroup{}
+
+	for _, fun := range discoveryFuncs {
+		wg.Add(1)
+
+		go func(f func() error) {
+			defer wg.Done()
+
+			err := f()
+			errCh <- err
+		}(fun)
+	}
+
+	wg.Wait()
+	close(errCh)
+
+	for e := range errCh {
+		if e != nil {
+			return e
+		}
+	}
+
+	dc.logger.Debug("cache has been updated")
 
 	return nil
 }
 
-func (j *discoveryCache) updateAPICache() error {
-	existingAPIs := make(map[string]bool)
-	apiServices, err := j.fetchAPIServices()
+func (dc *discoveryCache) handleAPISvc() error {
+	dc.logger.WithField("kind", "APIService").Trace("fetching API Services and updating cache")
+	query := map[string]string{
+		apic.FieldsKey: apiServerFields,
+	}
+	svcLink := mv1.NewAPIService("", dc.envName).GetKindLink()
+	url := dc.formatResourceURL(svcLink)
+
+	apiServices, err := dc.client.GetAPIV1ResourceInstancesWithPageSize(
+		query, url, apiServerPageSize,
+	)
 	if err != nil {
 		return err
 	}
 
 	for _, svc := range apiServices {
-		if j.migrator != nil {
+		if dc.migrator != nil {
 			var err error
-			svc, err = j.migrator.Migrate(svc)
+			svc, err = dc.migrator.Migrate(svc)
 			if err != nil {
 				return fmt.Errorf("failed to migrate service: %s", err)
 			}
 		}
 
-		externalAPIID, _ := util.GetAgentDetailsValue(svc, defs.AttrExternalAPIID)
-		// skip service without external api id
-		if externalAPIID == "" {
-			continue
-		}
-
-		agent.cacheManager.AddAPIService(svc)
-		primaryKey, _ := util.GetAgentDetailsValue(svc, defs.AttrExternalAPIPrimaryKey)
-		if primaryKey != "" {
-			existingAPIs[primaryKey] = true
-		} else {
-			existingAPIs[externalAPIID] = true
+		if err := dc.handleResource(svc); err != nil {
+			return err
 		}
 	}
 
 	return nil
 }
 
-func (j *discoveryCache) fetchAPIServices() ([]*apiV1.ResourceInstance, error) {
+func (dc *discoveryCache) handleServiceInstance() error {
+	dc.logger.WithField("kind", "APIServiceInstance").Trace("fetching API Service Instances and updating cache")
+	dc.instanceCacheLock.Lock()
+	defer dc.instanceCacheLock.Unlock()
+
 	query := map[string]string{
 		apic.FieldsKey: apiServerFields,
 	}
 
-	return GetCentralClient().GetAPIV1ResourceInstancesWithPageSize(
-		query, agent.cfg.GetServicesURL(), apiServerPageSize,
+	svcInstanceLink := mv1.NewAPIServiceInstance("", dc.envName).GetKindLink()
+	url := dc.formatResourceURL(svcInstanceLink)
+
+	serviceInstances, err := dc.client.GetAPIV1ResourceInstancesWithPageSize(
+		query, url, apiServerPageSize,
 	)
+	if err != nil {
+		e := utilErrors.Wrap(ErrUnableToGetAPIV1Resources, err.Error()).FormatError("APIServiceInstances")
+		dc.logger.Error(e)
+		return e
+	}
+
+	for _, instance := range serviceInstances {
+		if err := dc.handleResource(instance); err != nil {
+			return err
```

Comment:
@jcollins-axway for all of these funcs I am returning an error if the handler failed to process the resource. Maybe that isn't necessary and we should just log the error if that happens. I don't think everything needs to stop. What do you think?

Reply:
I agree, I think we need to look at these situations in the migration code as well. Was thinking about this after talking to Shane about the issue Dave hit.